lib/eig_trlm_3d.cpp

```diff
@@ -0,0 +1,701 @@
+    std::vector<ColorSpinorField> kSpace_t;
+
+    ColorSpinorParam csParamClone(kSpace[0]);
+    csParamClone.create = QUDA_ZERO_FIELD_CREATE;
+    csParamClone.change_dim(ortho_dim, 1);
+
+    for (int t = 0; t < ortho_dim_size; t++) {
+      if (!converged_3D[t]) {
+        int dim = n_kr - num_locked_3D[t];
+        int keep = iter_keep_3D[t];
+
+        ritz_mat_keep[t].resize(dim * keep);
+        for (int j = 0; j < dim; j++) {
+          for (int i = 0; i < keep; i++) { ritz_mat_keep[t][j * keep + i] = ritz_mat_3D[t][i * dim + j]; }
+        }
+
+        // Alias the vectors we wish to keep.
+        vector_ref<ColorSpinorField> vecs_locked(kSpace.begin() + num_locked_3D[t], kSpace.begin() + num_locked_3D[t] + dim);
+
+        // multiBLAS axpy. Create 3D vectors so that we may perform all t independent
+        // vector rotations.
+        vecs_t.reserve(dim);
+        kSpace_t.reserve(keep);
+
+        // Create 3D arrays
+        for (int i = vecs_t.size(); i < dim; i++) vecs_t.push_back(csParamClone);
+        for (int i = kSpace_t.size(); i < keep; i++) kSpace_t.push_back(csParamClone);
+
+	blas::zero(kSpace_t);
+
+        // Copy to data to 3D array, zero out workspace, make pointers
+        for (int i = 0; i < dim; i++) blas3d::copy(t, blas3d::COPY_TO_3D, vecs_t[i], vecs_locked[i]);
+
+        // Compute the axpy
+        blas::block::axpy(ritz_mat_keep[t], {vecs_t.begin(), vecs_t.begin() + dim},
+                          {kSpace_t.begin(), kSpace_t.begin() + keep});
+
+        // Copy back to the 4D workspace array
+
+        // Copy compressed Krylov
+        for (int i = 0; i < keep; i++) {
+          blas3d::copy(t, blas3d::COPY_FROM_3D, kSpace_t[i], kSpace[num_locked_3D[t] + i]);
+        }
+
+        // Update residual vector
+        blas3d::copy(t, blas3d::COPY_TO_3D, vecs_t[0], kSpace[n_kr]);
+        blas3d::copy(t, blas3d::COPY_FROM_3D, vecs_t[0], kSpace[num_locked_3D[t] + keep]);
+
+        // Update sub arrow matrix
+        for (int i = 0; i < keep; i++)
+          beta_3D[t][i + num_locked_3D[t]] = beta_3D[t][n_kr - 1] * ritz_mat_3D[t][dim * (i + 1) - 1];
+      }
+    }
+
+    getProfile().TPSTOP(QUDA_PROFILE_COMPUTE);
+  }
+
+
+  // Orthogonalise r[t][0:] against V_[t][0:j]
+  void TRLM3D::blockOrthogonalize3D(std::vector<ColorSpinorField> &vecs, std::vector<ColorSpinorField> &rvecs, int j)
+  {
+    for (int i = 0; i < j; i++) {
+      std::vector<Complex> s_t(ortho_dim_size, 0.0);
+      std::vector<Complex> unit_new(ortho_dim_size, {1.0, 0.0});
+
+      // Block dot products stored in s_t.
+      blas3d::cDotProduct(s_t, vecs[i], rvecs[0]);
+      for (int t = 0; t < ortho_dim_size; t++) {
+        s_t[t] *= active_3D[t] ? -1.0 : 0.0;
+      }
+
+      // Block orthogonalise
+      blas3d::caxpby(s_t, vecs[i], unit_new, rvecs[0]);
+    }
+  }
+
+  void TRLM3D::prepareInitialGuess3D(std::vector<ColorSpinorField> &kSpace, int ortho_dim_size)
+  {
+    if (sqrt(blas::norm2(kSpace[0])) == 0.0) {
```

Comment:
I learned the hard way through interfacing with MILC that we should also check for +/- inf and nan in the context of completely uninitialized data, see here: https://github.com/lattice/quda/blob/develop/lib/eigensolve_quda.cpp#L145-L150

Reply:
Good call.  Will use `std::isfinite` here.

Reply:
c5b568171